Flutter_TDD_Architecture_Course_notes.md

```diff
@@ -277,3 +277,740 @@ class GetRandomNumberTrivia extends UseCase<NumberTrivia, NoParams> {
+import '../datasources/number_trivia_remote_data_source.dart';
+
+class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
+  final NumberTriviaRemoteDataSource remoteDataSource;
+  final NumberTriviaLocalDataSource localDataSource;
+  final NetworkInfo networkInfo;
+
+  NumberTriviaRepositoryImpl({
+    @required this.remoteDataSource,
+    @required this.localDataSource,
+    @required this.networkInfo,
+  });
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
+    // TODO: implement getConcreteNumberTrivia
+    return null;
+  }
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
+    // TODO: implement getRandomNumberTrivia
+    return null;
+  }
+}
+```
+#### 6. Repository Implementation
+> Video 6: https://www.youtube.com/watch?v=bfEKPKKy9dA
+
+In the `sixth` tutorial start by explaining that we will have to know if the user has an internet connection and for that a test will be done.
+
+#### Device Is Online
+
+```dart
+group('getConcreteNumberTrivia', () {
+  // DATA FOR THE MOCKS AND ASSERTIONS
+  // We'll use these three variables throughout all the tests
+  final tNumber = 1;
+  final tNumberTriviaModel =
+      NumberTriviaModel(number: tNumber, text: 'test trivia');
+  final NumberTrivia tNumberTrivia = tNumberTriviaModel;
+
+  test('should check if the device is online', () {
+    //arrange
+    when(mockNetworkInfo.isConnected).thenAnswer((_) async => true);
+    // act
+    repository.getConcreteNumberTrivia(tNumber);
+    // assert
+    verify(mockNetworkInfo.isConnected);
+  });
+});
+```
+The test will `fail` because we have to add features.
+Add the code to the `number_trivia_repository_impl_test.`dart.
+
+```dart
+group('device is online', () {
+  // This setUp applies only to the 'device is online' group
+  setUp(() {
+    when(mockNetworkInfo.isConnected).thenAnswer((_) async => true);
+  });
+
+  test(
+    'should return remote data when the call to remote data source is successful',
+    () async {
+      // arrange
+      when(mockRemoteDataSource.getConcreteNumberTrivia(tNumber))
+          .thenAnswer((_) async => tNumberTriviaModel);
+      // act
+      final result = await repository.getConcreteNumberTrivia(tNumber);
+      // assert
+      verify(mockRemoteDataSource.getConcreteNumberTrivia(tNumber));
+      expect(result, equals(Right(tNumberTrivia)));
+    },
+  );
+});
+```
+Now we need to do some tests and add some features.
+
+```dart
+test(
+  'should cache the data locally when the call to remote data source is successful',
+  () async {
+    // arrange
+    when(mockRemoteDataSource.getConcreteNumberTrivia(tNumber))
+        .thenAnswer((_) async => tNumberTriviaModel);
+    // act
+    await repository.getConcreteNumberTrivia(tNumber);
+    // assert
+    verify(mockRemoteDataSource.getConcreteNumberTrivia(tNumber));
+    verify(mockLocalDataSource.cacheNumberTrivia(tNumberTrivia));
+  },
+);
+```
+
+```dart
+impl.dart
```

Comment:
`impl.dart` is a _file_ not a line of code. Why is it in this code block?

Reply:
@nelsonic It was a mistake when I was writing the name of the file he passed down.